core/src/worker/activities.rs

```diff
@@ -509,6 +527,53 @@ where
                                     tt,
                                     ActivityCancelReason::WorkerShutdown,
                                 ));
+                            } else {
+                                // Fire off task to keep track of local timeouts. We do this so that
+                                // activities can still get cleaned up even if the user isn't
+                                // heartbeating.
+                                let local_timeout_buffer = self.local_timeout_buffer;
+                                let sched = task
+                                    .resp
+                                    .schedule_to_close_timeout
+                                    .clone()
+                                    .and_then(|to| {
+                                        // Must filter out zero before any subtraction math here
+                                        if to.seconds == 0 && to.nanos == 0 {
+                                            None
+                                        } else {
+                                            Some(to)
+                                        }
+                                    })
+                                    .zip(task.resp.scheduled_time.clone())
+                                    .and_then(|(to, st)| {
+                                        let now = SystemTime::now();
```

Comment:
I'm scared of SystemTime and it not being monotonic. Should we use `Instant` here?

Reply:
We can't. System time is the only thing the protobuf stuff can be converted into. The unwrap_or / defaults provide protection against stuff not making sense